File: Quantitative_Trading_Chan/example7_4.m
clear;

lookback=252; % use lookback days as estimation (training) period for determining factor exposures.
numFactors=5;
topN=50; % for trading strategy, long stocks with topN expected 1-day returns.

% test on SP600 smallcap stocks. (This MATLAB binary input file
% contains tday, stocks, op, hi, lo, cl arrays.
load('IJR_20080114');

mycls=fillMissingData(cl);

positionsTable=zeros(size(cl));

% note the rows of dailyret are the observations at different time
% periods
dailyret=(mycls-lag1(mycls))./lag1(mycls);

for t=lookback+1:length(tday)

    % here the columns of R are the different observations.
    R=dailyret(t-lookback+1:t,:)'; 

    % avoid any stocks with missing returns
    hasData=find(all(isfinite(R), 2)); 
    
    R=R(hasData, :);
    
    avgR=smartmean(R, 2);
    % subtract mean from returns
    R=R-repmat(avgR, [1 size(R, 2)]); 

    % compute covariance matrix, with observations in rows.
    covR=smartcov(R'); 

    % X is the factor exposures matrix, B the variances of factor returns
    [X, B]=eig(covR); 
    
    X(:, 1:size(X, 2)-numFactors)=[]; % Retain only numFactors

    % b are the factor returns for time period t-1 to t.
    results=ols(R(:, end), X); 
    b=results.beta;

    % Rexp is the expected return for next period assuming factor
    % returns remain constant.
    Rexp=avgR+X*b;
    
    [foo idxSort]=sort(Rexp, 'ascend');
    
    positionsTable(t, hasData(idxSort(1:topN)))=-1; % short topN stocks with lowest expected returns
    positionsTable(t, hasData(idxSort(end-topN+1:end)))=1; % buy topN stocks with highest  expected returns
end

ret=smartsum(backshift(1, positionsTable).*dailyret, 2); % compute daily returns of trading strategy
avgret=smartmean(ret)*252 % compute annualized average return of trading strategy
% A very poor return!
% avgret =
% 
%    -1.8099

File: Quantitative_Trading_Chan/python/example7_4.py
# coding: utf-8

# Principal Component Analysis as an Example of Factor Model

import numpy as np
import pandas as pd
from numpy.linalg import eig
from statsmodels.api import OLS

lookback=252 # training period for factor exposure
numFactors=5
topN=50 # for trading strategy, long stocks with topN exepcted 1-day returns

df=pd.read_table('IJR_20080114.txt')
df['Date']=df['Date'].astype('int')
df.set_index('Date', inplace=True)
df.sort_index(inplace=True)
df.fillna(method='ffill', inplace=True)

dailyret=df.pct_change() # note the rows of dailyret are the observations at different time periods
positionsTable=np.zeros(df.shape)

for t in np.arange(lookback,df.shape[0]): 
    R=dailyret.iloc[t-lookback+1:t+1,].T # here the columns of R are the different observations.
    hasData=np.where(R.notna().all(axis=1))[0]
    R.dropna(inplace=True) # avoid any stocks with missing returns
    avgR=R.mean(axis=1)
    R=R.values-avgR.values.reshape((R.shape[0], 1)) # subtract mean from returns
    covR=pd.DataFrame(R.T).cov() # compute covariance matrix, with observations in rows.
    B, X=eig(covR) # X is the factor exposures matrix, B the variances of factor returns
    X=X[:, 0:numFactors] # Retain only numFactors
    model=OLS(R[:, -1], X)
    results=model.fit()
    b=results.params    # b are the factor returns for time period t-1 to t.
    Rexp=avgR+np.dot(X,b) # Rexp is the expected return for next period assuming factor returns remain constant.
    idxSort=Rexp.argsort()  
    
    positionsTable[t, hasData[idxSort.values[np.arange(0, topN)]]]=-1
    positionsTable[t, hasData[idxSort.values[np.arange(-topN,0)]]]=1
    

capital=np.nansum(np.array(abs(positionsTable).shift()), axis=1)
positionsTable[capital==0,]=0
capital[capital==0]=1
ret=np.nansum(np.array(pd.DataFrame(positionsTable).shift())*np.array(dailyret), axis=1)/capital
avgret=np.nanmean(ret)*252
